### PR TITLE
Added subscriber notifications

### PR DESCRIPTION
**Pull request by umbreak**
- Added a way to subscribe to the key value store events.
- Added `remove` method to delete a key from the key value store map.

### CODE REVIEW

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    case _ => // drop
+  }
+}
+
+object KeyValueStoreSubscriber {
+
+  /**
+    * Enumeration of types related to changes to the key value store.
+    */
+  sealed trait KeyValeStoreChange extends Product with Serializable
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala]: typo

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import ch.epfl.bluebrain.nexus.service.indexer.cache.KeyValueStoreSubscriber._
+
+trait OnKeyValueStoreChange[K, V] {
+
+  /**
+    * Method that gets triggered when a change to key value store occurs.
+    *
+    * @param value the changes made
+    */
+  def apply(value: KeyValeStoreChanges[K, V]): Unit
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala]: I'm wondering if we should make this take an `F[_]: Async` as most of the time I suspect we'd want to do something with the changes and our API is almost all defined like this... thoughts?

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+object OnKeyValueStoreChange {
+  def noEffect[K, V]: OnKeyValueStoreChange[K, V] = _ => ()
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala:27]: docs

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  /**
+    * Constructs the [[KeyValueStoreSubscriber]] actor.
+    *
+    * @param id       the ddata key
+    * @param onChange the method that gets triggered whenever a change to the key value store occurs
+    * @tparam K the key type
+    * @tparam V the value type
+    * @return an [[ActorRef]] of the [[KeyValueStoreSubscriber]] actor
+    */
+  final def apply[K, V](id: String, onChange: OnKeyValueStoreChange[K, V])(implicit as: ActorSystem): ActorRef =
```
Comment (bogdanromanx) [inline modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala:119]: the fact that the actor id is equal to the map key it means we can only have a single actor per cache. we could instead pass the id to the actor constructor, let it subscribe by itself and leave the actor id to be automatically generated.

Comment (bogdanromanx) [reply modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala:119]: this should allow us to have multiple subscriptions on the store api, something like:
`def subscribe(f: KeyValueStoreChanges[K, V] => ()): F[Subscription] = ???`

and
```
class Subscription[F[_]] private(key: String, ref: ActorRef) {
  def unsubscribe: F[Unit] = ???
}
```

**Review by bogdanromanx — CHANGES_REQUESTED** [review]

**Comment by bogdanromanx** [comment]
This is pretty cool, BTW...

File: modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala (hunk 4 above)
Comment (umbreak) [reply modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala:119]: One actor per cache makes sense and it is more than enough. We are not planning on having 300 caches.

**Review by bogdanromanx — APPROVED** [review]

**Pull request merged**